WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCardViewModel.swift

```diff
@@ -66,9 +67,27 @@ final class MostActiveCouponsCardViewModel: ObservableObject {
     func reloadData() async {
         syncingData = true
         syncingError = nil
+        rows = []
+        resultsController = nil
+
         do {
-            let activeCouponReports = try await mostActiveCoupons()
-            coupons = try await loadCouponDetails(for: activeCouponReports)
+            let couponReports = try await fetchMostActiveCoupons()
+
+            if couponReports.isNotEmpty {
+                /// Load and display coupons from storage
+                configureResultsController(for: couponReports)
+
+                /// If all or some coupons are not available in storage
+                /// then fetch from remote
+                if rows.count != couponReports.count {
+                    try await synchronizeCoupons(for: couponReports)
+                } else {
+                    /// Refresh coupons in background
+                    Task { @MainActor in
+                        try await synchronizeCoupons(for: couponReports)
+                    }
+                }
```

Comment:
I found a possible edge case with my test store. The coupon reports return 3 items and one of them has already been removed so the sync call cannot find that one coupon => the code always fall to the `else` case. I believe this is not a serious case, but that makes me wonder how often this happens and if it makes our storage usage become a waste. I'll leave this for you to decide if we should go ahead without storage (sorry for the back and forth 🙇).

Reply:
Nice find, Huong! I will keep the storage-related logic as it doesn't create any issues.